### ISSUE TITLE
can't guarantee the reproducibility with seeding everything

### ISSUE THREAD

**Issue by squirrel16** [jwyang/faster-rcnn.pytorch#394]
In the training code, I set all the random seeds,
```
  random.seed(cfg.RNG_SEED)
  np.random.seed(cfg.RNG_SEED)
  torch.manual_seed(cfg.RNG_SEED)
  if args.cuda:
    torch.cuda.manual_seed(cfg.RNG_SEED)
    torch.cuda.manual_seed_all(cfg.RNG_SEED)
  torch.backends.cudnn.deterministic = True
  torch.backends.cudnn.benchmark = False
  os.environ['PYTHONHASHSEED'] = str(cfg.RNG_SEED)
```
But when I tried to run the same code (with same settings and hyperparameters) multiple times, I got different loss curves and evaluation results.
Do you have any idea about this? Very much thanks.@jwyang

**Comment by Rabbott93** [comment]
@squirrel16 did you ever solve this issue? I am having the same problem

**Comment by Jokoe66** [comment]
@squirrel16 I have found the same problem. Have you got any ideas?

**Comment by Jokoe66** [comment]
> @squirrel16 I have found the same problem. Have you got any ideas?

@jwyang @squirrel16 Finally I find that with RoiPooling the model is reproducible after every rng_seed is set and cudnn is deterministic. I' m not clear what makes RoIAlign non-deterministic. I' m grateful if you can dig out the reason.

**Comment by yelantingfeng** [comment]
Atomic add  in backward implementation.

**Comment by mohanrobotics** [comment]
@Jokoe66 Can you please where did you add the seeding ?  I added the below code in training file and changed the pooling mode to 'pool' in the config file. Still not able to reproduce the results.
`  os.environ['PYTHONHASHSEED'] = str(cfg.RNG_SEED)
random.seed(cfg.RNG_SEED)
  np.random.seed(cfg.RNG_SEED)
  torch.manual_seed(cfg.RNG_SEED)
  if args.cuda:
    torch.cuda.manual_seed(cfg.RNG_SEED)
    torch.cuda.manual_seed_all(cfg.RNG_SEED)
  torch.backends.cudnn.deterministic = True
  torch.backends.cudnn.benchmark = False
`

**Comment by Jokoe66** [comment]
I added the seeding code behind pprint.pprint(cfg).




------------------ Original ------------------

**Comment by mohanrobotics** [comment]
I still not able to reproduce. May be because of multi-gpu I guess. Did anyone able to reproduce results using mulitgpu?

**Comment by Jokoe66** [comment]
Btw, did you set seed in worker_init_fn used in dataloader? It should be done when num_workers is greater than zero, otherwise the results are still irreproducible. ------------------ 原始邮件 ------------------
发件人: "Mohandass Muthuraja"<notifications@github.com>
发送时间: 2019年10月17日(星期四) 凌晨2:13
收件人: "jwyang/faster-rcnn.pytorch"<faster-rcnn.pytorch@noreply.github.com>;
抄送: "Jokoe66"<531211903@qq.com>;"Mention"<mention@noreply.github.com>;
主题: Re: [jwyang/faster-rcnn.pytorch] can't guarantee the reproducibilitywith seeding everything (#394)



I still not able to reproduce. May be because of multi-gpu I guess. Did anyone able to reproduce results using mulitgpu?
 
&mdash;

**Comment by mohanrobotics** [comment]
yes . I did . 
```def _init_fn(worker_id):
    np.random.seed(3)```
and 
`    dataloader = torch.utils.data.DataLoader(dataset, batch_size=args.batch_size,
                                             sampler=sampler_batch, num_workers=args.num_workers,
                                             worker_init_fn=_init_fn)`

**Comment by Jokoe66** [comment]
Usually the seed in worker_init_fn is set to SEED+worker_id. Maybe you can have a try.


------------------ Original ------------------
